### PR TITLE
🔏 feat: add csrf protection

### LINKED ISSUE #59: CSRF protection
A middleware based approach (using https://github.com/pillarjs/csrf) exposing the CSRF token to the templating layer through a global property or helper.

### PR DESCRIPTION
**Pull request by adamkdean**
This introduces CSRF protection.

Closes #59.

# About

Cross-Site Request Forgery (CSRF) is an attack that forces an end user to execute unwanted actions on a web application in which they're currently authenticated. CSRF attacks specifically target state-changing requests, not theft of data, since the attacker has no way to see the response to the forged request. [More about CSRF](https://github.com/pillarjs/understanding-csrf).

This feature allows developers to use a per-request CSRF token which will be injected into the view model, and ensures that all POST requests supply a correct CSRF token. Without a correct token, with CSRF enabled, users will be greeted with a 403.

This feature uses [csurf](https://github.com/expressjs/csurf).

# Usage

To enable CSRF, set the `security.csrf` config option in your `config/config.{env}.json` file:

```
"security": {
  "csrf": true
}
```

Once enabled, the variable `csrfToken` will be injected into the viewModel. You will need to add this to any forms which perform a `POST` using the field name `_csrf`, like so:

```
<form action="/" method="post">
  <input type="text" name="test_input_safe">
  <input type="hidden" name="_csrf" value="{csrfToken}">
  <input type="submit" value="Submit form">
</form>
```

If the CSRF token provided is incorrect, or one isn't provided, then a `403 forbidden` error will occur.

A working example can be found here: [dadi-web-csrf-test](https://github.com/adamkdean/dadi-web-csrf-test).

### CODE REVIEW

**Comment by adamkdean** [comment]
@jimlambie do you know why these tests failed? Am I a bad person or is it an unrelated issue?

**Comment by mingard** [comment]
@adamkdean something funny going on with the doc param in the config schema e.g.
```json
  csrf: {
      doc:
        'If true, a CSRF token will be provided, and all form submissions must include this as _csrf',
      format: '*',
      default: false
    }
```

**Comment by adamkdean** [comment]
@mingard oh wow, just seen this now in the diff, that's weird, looks like something has changed all the docs? I'll fix it tomorrow and tidy it up. Good spot!

**Comment by jimlambie** [comment]
@abovebored can we disable the prettifying of the config.js file?

**Comment by jimlambie** [comment]
Actually, @adamkdean, I think the issue is due to an incorrect version of `sinon` being installed. Please change in the package file:

from `"sinon": "latest",`
to `"sinon": "2.x.x",`

**Comment by abovebored** [comment]
@adamkdean no need to be alarmed, it's [prettier](https://github.com/dadi/web/blob/master/package.json#L18) (trimming the line length)

**Comment by adamkdean** [comment]
@abovebored @jimlambie `ready for review`

**Comment by adamkdean** [comment]
@jimlambie my understanding of csurf's implementation is that if it's used (csrf enabled in web config), then any `POST` requests must provide a valid token under the name `_csrf`, and if it either isn't present or is incorrect, then a 403 is generated, which currently gets picked up by web for the custom error page. If you could knock up a couple of tests for this, that'd be great!

**Comment by jimlambie** [comment]
@adamkdean I have added 4 tests in test/unit/controller.js, however the final test (HTTP 200 with valid token) is failing. Could you please have a look, as perhaps I misunderstand the implementation.

Adding the following to `node_modules/csrf/index.js` L142 gives me different token values:

`console.log(token, expected)`

**Comment by adamkdean** [comment]
FYI the issue was that the secret is stored in a cookie, and needs to be passed back to the host with the next request, though if sessions are enabled, they are used instead.

**Comment by adamkdean** [comment]
This time we're gonna get ✅ , I can feel it.

File: dadi/lib/index.js [hunk 1]
```diff
@@ -17,6 +17,8 @@ var raven = require('raven')
 var serveFavicon = require('serve-favicon')
 var serveStatic = require('serve-static')
 var session = require('express-session')
+var csrf = require('csurf')
```
Comment (eduardoboucas) [inline dadi/lib/index.js]: Bonus points (and a beer) are granted if you replace `var` with `const/let` as you go. ⭐️ 🍺

**Comment by eduardoboucas** [comment]
> We've enabled this by default and also made a change so the error stack (for this and all other errors) are hidden when the environment is set to production mode 🕵🏻

Does making this enabled by default introduce a breaking change?

**Comment by jimlambie** [comment]
Web 4.0...

**Comment by eduardoboucas** [comment]
If I understand correctly, if this is disabled in the config then everything works exactly as before, right? If that's the case, perhaps we can introduce the feature but make it disabled by default and revisit when we release the next major.

File: dadi/lib/index.js (hunk 1 above)
Comment (adamkdean) [reply dadi/lib/index.js]: @eduardoboucas I wasn't sure if we were doing that or if the plan was to stick with the existing style. Maybe it'd be worth tidying up the dependencies at least.

**Comment by adamkdean** [comment]
@eduardoboucas yes, you're correct. When disabled, everything will work as expected, but if enabled by default, all existing forms will break. Definitely requires a major bump.

**Pull request merged**